posts/2021-06-11-kubernetes-operator-with-java-and-quarkus.adoc

```diff
@@ -0,0 +1,712 @@
+               .addToLabels("app", spec.getName())
+               .endMetadata()
+               .withNewSpec()
+               .addNewContainer()
+               .withName(spec.getName())
+               .withImage(spec.getImage())
+               .withCommand("sh", "-c", "while sleep 5; do cat /tmp/data.txt; done")
+               .addNewPort()
+               .withContainerPort(80)
+               .endPort()
+               .endContainer()
+               .endSpec()
+               .endTemplate()
+               .withNewSelector()
+               .addToMatchLabels("app", spec.getName())
+               .endSelector()
+               .endSpec()
+               .build();
+
+       kubernetesClient.apps().statefulSets().inNamespace(namespace).create(statefulset);
+
+       return UpdateControl.noUpdate();
+
+   } else if (statefulset.getStatus().getReplicas() < spec.getReplicas()) {
+
+       log.info("Scale statefulset up: " + spec.getName());
+
+       kubernetesClient.apps()
+               .statefulSets()
+               .inNamespace(namespace).withName(spec.getName()).scale(statefulset.getStatus().getReplicas() + 1, true);
+
+       return UpdateControl.noUpdate();
+
+   } else if (statefulset.getStatus().getReplicas() > spec.getReplicas()) {
+
+       log.info("Scale statefulset down: " + spec.getName());
+
+       kubernetesClient.apps()
+               .statefulSets()
+               .inNamespace(namespace).withName(spec.getName()).scale(spec.getReplicas(), true);
+
+       return UpdateControl.noUpdate();
+   }
+
+   return UpdateControl.noUpdate();
+----
+====
+
+Containers are running following command to check if the data is present in the container, and outputs an error to a log if data is not present.
+
+----
+"while sleep 5; do cat /tmp/data.txt; done"
+----
+
+PodWatcher watches pod creatiion and modification
+
+[,java]
+----
+public class PodWatcher {
+
+private final KubernetesClient kubernetesClient;
+
+void onStartup(@Observes StartupEvent startupEvent) throws IOException {
+
+   public PodWatcher(KubernetesClient kubernetesClient) {
+      this.kubernetesClient = kubernetesClient;
+   }
+
+
+   kubernetesClient.pods().watch(new Watcher<Pod>() {
+
+   @Override
+   public void eventReceived(Action action, Pod pod) {...}
+
+   @Override
+   public void onClose(WatcherException e) {...}
+});
+
+}
+----
+
+Start an operator
+
+.QuarkusApplication.java
+[,java]
+----
+@QuarkusMain
+public class Main implements QuarkusApplication {
+
+   @Inject
+   Operator operator;
+
+   public static void main(String... args) {
+       Quarkus.run(Main.class, args);
+   }
+
+   @Override
+   public int run(String... args) throws Exception {
+       operator.start();
+       Quarkus.waitForExit();
+       return 0;
+   }
+}
+----
+
+When pod is created, operator fetches logs and
+checks if the data from custom object is present in the pod, in case if it is not present operator outputs data to a file in the pod.
```

Comment:
```suggestion
checks if the data from the custom object is present in the pod, in case if it is not present operator outputs data to a file in the pod.
```